src/mpid/ch4/src/ch4_progress.c

```diff
@@ -30,16 +30,11 @@ int MPIDI_Progress_test(int flags)
     if (flags & MPIDI_PROGRESS_HOOKS) {
         for (i = 0; i < MPIDI_global.registered_progress_hooks; i++) {
             progress_func_ptr_t func_ptr = NULL;
-            MPID_THREAD_CS_ENTER(VCI, MPIDIU_THREAD_PROGRESS_HOOK_MUTEX);
```

Comment:
I missed the discussion. I understand the direction is to make `.active` atomic; but how are these mutexes "incorrect"?

Reply:
They were not being used by the progress engine, where the variable is actually checked.  So we'd need to add those mutexes to the progress engine too to make it correct.  Instead, changing everything to atomics seemed like the right approach because on most platforms load/store atomics are converted to simple load/store operations, so there's no overhead.

A second incorrect behavior, that neither the mutex code nor the atomic code fixes, is that there's no lock acquired for the "if schedule is empty, deactivate hook" path.  So one thread can check enqueue a schedule, but another thread might deactivate the hook even when the schedule is not empty.  That requires a sched-specific mutex, which I believe you are working on in a separate PR.

Reply:
> That requires a sched-specific mutex, which I believe you are working on in a separate PR.

Merged already.

Reply:
@rzambre I talked to Hui and he said that you are working on using these mutexes correctly?  Do you have a PR?  The current code has a race condition.

Reply:
@rzambre Nevermind, after one more discussion with Hui, we think the code is correct.  Sorry for the noise.

Reply:
Since the question will pop up again, maybe we should add a comment explain it.